x-pack/plugins/fleet/server/plugin.ts

```diff
@@ -362,6 +367,8 @@ export class FleetPlugin
             .getSavedObjects()
             .getScopedClient(request, { excludedExtensions: [SECURITY_EXTENSION_ID] });
 
+        appContextService.setAuditLogger(request);
```

Comment:
Not sure this will work, we can have concurrent request and `appContextService` is a global state shared between all of this no?

Reply:
Hmm you might be right. We only call `appContextService.start` during the Fleet plugin's `start` lifecycle, which I'm not sure is called per-request. 

https://github.com/elastic/kibana/blob/adb0bb58d3dae83d761a03edc5ac77793c8cafe7/x-pack/plugins/fleet/server/plugin.ts#L436

I'll do some testing and see if I can validate this.

Reply:
I think we may not have other solution than passing another argument with a context to all the services where we want to have audit logging

Reply:
Yeah I think this needs to move to its own service outside of `appContextService` - we're overriding the global context's `auditLogger` value every time a new request comes in. Should be relatively straightforward to set up.

Reply:
> I think we may not have other solution than passing another argument with a context to all the services where we want to have audit logging

The problem with this is that most of our services are singletons, meaning we don't have a `constructor` where we can pass something like `auditLoggingService` as an argument. We'd have to update every single static method on these classes with a new argument which seems very cumbersome.

Reply:
What about deprecating appContextService and having a new context container that could be scopped to a request or use in background, it's a big refacto but hopefully we will have to do it once if we need to add other context services

in an handler
```
const scoppedContextService = new ScopedContextService(request);

const { items, total, page, perPage } = await agentPolicyService.list(soClient, {
  withPackagePolicies,
  ...restOfQuery,
}, scoppedContextService); // <----
```

In a background task 
```
const contextService = new ContextService(request);

const { items, total, page, perPage } = await agentPolicyService.list(soClient, {
  withPackagePolicies,
  ...restOfQuery,
}, contextService); // <----
```

Reply:
I'm definitely +1 on deprecating `AppContextService` overall in Fleet, as it adds a lot of usability issues to Fleet's server side code. However, that seems like too big of an undertaking to commit to in this PR. 

`AsyncLocalStorage` could make a lot of sense for our use case, and it seems like Kibana already uses it for its `ExecutionContextService` class here:

https://github.com/elastic/kibana/blob/main/packages/core/execution-context/core-execution-context-server-internal/src/execution_context_service.ts

I believe `AsyncLocalStorage` was considered stable in Node 14, so we should be okay here in terms of support. Maybe something worth exploring here.